### PR TITLE
HOTFIX: Fix recent protocol breakage

### PR DESCRIPTION
**Pull request by hachikuji**
KIP-345 and KIP-392 introduced a couple breaking changes for old versions of bumped protocols. This patch fixes them.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by rajinisivaram** [comment]
retest this please

**Review by abbccdda — APPROVED** [review]
LGTM. Do we also want to add some documentation to make sure we use the null type correctly?

**Comment by cmccabe** [comment]
LGTM

**Comment by guozhangwang** [comment]
LGTM, waiting for some more unit tests to be added before merging.

**Comment by cmccabe** [comment]
> Do we also want to add some documentation to make sure we use the null type correctly?

There is some documentation here: https://github.com/apache/kafka/tree/trunk/clients/src/main/resources/common/message
Let me know if you have any ideas for improving it.

**Review by noslowerdna — APPROVED** [review]

**Comment by guozhangwang** [comment]
I think the docs around nullable default values can be further improved. To me at least, there are two conflicting semantics it seems:

1. default applies ONLY to non-existing versions, it is what you get f the message didn't have the field

2. as stated in the docs, `you can only specify null as a default if all versions of the field are nullable.` so it seems default values are indeed correlated to the nullable versions as well, and they will be used if the field is not specified.

File: clients/src/test/java/org/apache/kafka/common/message/MessageTest.java
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/common/message/MessageTest.java]: The last parameter seems not correct? Should this be a `testMessageRoundTripsFromVersion` call?

**Review by guozhangwang — COMMENTED** [review]
I have a question about one line in the added unit test, otherwise lgtm.

File: clients/src/test/java/org/apache/kafka/common/message/MessageTest.java
Comment (hachikuji) [reply clients/src/test/java/org/apache/kafka/common/message/MessageTest.java]: Yes, it should be. Thanks.

**Comment by guozhangwang** [comment]
LGTM. I will merge after green builds.

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

**Comment by abbccdda** [comment]
Verified the system test is working: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2617/console

**Comment by hachikuji** [comment]
Merging to trunk and 2.3. I'll file a JIRA for the failing test if one doesn't exist already, but it looks unrelated.

**Comment by guozhangwang** [comment]
retest this please

**Pull request merged**
